### ISSUE TITLE
Audit system integrity

### ISSUE THREAD

**Issue by SaswatPadhi** [QubesOS/qubes-issues#5335]
**The problem you're addressing (if any)**
I think it would great to have some sort of infrastructure to detect (automatically, or at least when requested by the user) if a Qubes system has been compromised. One of the things that Qubes actively discourages and makes it hard for users to do, is modifying dom0. I really like that aspect of Qubes, but the system doesn't provide any means of detecting intrusion.

**Describe the solution you'd like**
Windows systems have a feature called "System File Checker (SFC)" that scans the OS-specific files for any unexpected modifications. At minimum, Qubes should be some script that could check if dom0 has been somehow modified, of course excluding logs and "normal modifications" like those. A stretch goal could be to also check integrity of Qubes scripts on the TemplateVMs -- because the VMs communicate to dom0 only via those scripts.

**Where is the value to a user, and who might that user be?**
This would protect users from having a false sense of security on compromised systems.

**Comment by marmarek** [comment]
>  This would protect users from having a false sense of security on compromised systems.

This would have exactly opposite effect. All kind of mechanisms that say "your system is clean for sure" will make more harm when the system actually isn't. And it is [very hard/impossible](https://www.qubes-os.org/news/2017/04/26/qubes-compromise-recovery/#digression-about-detecting-compromises) to make such detection reliable (see also [here](https://www.qubes-os.org/intro/#arent-antivirus-programs-and-firewalls-enough)). Furthermore, the first thing successful attacker would do, is to replace this mechanism with something that always says it's clean. Depending on actual mechanism it may be easier or harder, may require different degree of compromise (dom0 user, dom0 kernel, hypervisor, firmware etc), but still possible. Don't forget also that not every compromise would result in some files modified - it's enough to compromise a running process/kernel/hypervisor to do harm.

On the other hand, having trust in some some system parts, you can verify others starting from there. For example https://github.com/tasket/Qubes-VM-hardening implements this at TemplateVM->AppVM level. Or [Anti Evil Maid](https://www.qubes-os.org/doc/anti-evil-maid/) does it on CPU(+parts of BIOS) -> Xen+Kernel level. But you need that _other_ part you trust. Dom0 cannot reliably verify itself.

Anyway, you can easily check if dom0 files are as they were installed using package manager, specifically with `rpm -Va`. But as said previously, if someone manages to compromise dom0, she/he could easily circumvent any detection mechanism running in dom0. There are various ways to make it even harder than it is now, for example making most of dom0 filesystem read only and signed (something that would be more realistic after moving GUI out of dom0). 

In short: making it harder to break into dom0 and/or persist there -> yes, definitely. Adding "this system is clean" indicator -> no.

I'm closing this for now, because different hardening mechanisms are already covered in documentation and other tickets.

**Comment by SaswatPadhi** [comment]
Thank you for the very detailed answer! Appreciate it.
